docs: introduce docs for clustering (#4344)
grafana/agent: docs/sources/flow/reference/cli/run.md

[component controller]: {{< relref "../../concepts/component_controller.md" >}}

> ## Clustering (beta)

Reviewer: Heads up: there should be a changelog entry indicating that clustering in Flow mode has graduated to beta.

[component controller]: {{< relref "../../concepts/component_controller.md" >}}

## Clustering (beta)

The `--cluster.enabled` command-line argument starts Grafana Agent in
[clustering][] mode. The rest of the `--cluster.*` command-line flags can be
used to configure how nodes discover and connect to one another.

Each cluster member’s name must be unique within the cluster. Nodes which try
to join with a conflicting name are rejected and will fall back to
bootstrapping a new cluster of their own.

Peers communicate over HTTP/2 on the agent's built-in HTTP server. Each node
must be configured to accept connections on `--server.http.listen-addr` and the
address defined or inferred in `--cluster.advertise-address`.

If the `--cluster.advertise-address` flag is not explicitly set, the agent

> tries to infer a suitable one from the `eth0` and `en0` local network

Reviewer 2: `eth0` and `en0` are Linux-specific terms. Maybe someone will think it doesn't work on windows?

Reviewer: That's correct, eth0 and en0 are always used as the defaults meaning it won't work on Windows. I've clarified that Windows users need to provide a value for this flag here.

## Clustering (beta)

The `--cluster.enabled` command-line argument starts Grafana Agent in
[clustering][] mode. The rest of the `--cluster.*` command-line flags can be
used to configure how nodes discover and connect to one another.

Each cluster member’s name must be unique within the cluster. Nodes which try
to join with a conflicting name are rejected and will fall back to
bootstrapping a new cluster of their own.

Peers communicate over HTTP/2 on the agent's built-in HTTP server. Each node
must be configured to accept connections on `--server.http.listen-addr` and the
address defined or inferred in `--cluster.advertise-address`.

If the `--cluster.advertise-address` flag is not explicitly set, the agent
tries to infer a suitable one from the `eth0` and `en0` local network
interfaces. If the advertised address cannot be determined, the agent will
fail to start. Windows users must explicitly pass a value for
`--cluster.advertise-address`, since Windows does not have interfaces name
`eth0` or `en0`.

The comma-separated list of addresses provided in `--cluster.join-addresses`
can either be IP addresses with an optional port, or DNS records to lookup.

> The ports on the this list of addreses default to the port used for the HTTP

Reviewer 3: ```suggestion
The ports on the list of addresses default to the port used for the HTTP
```

> listener if not explicitly provided. It’s generally recommended to
> align the port numbers on as many nodes as possible to simplify the deployment

Reviewer 3: ```suggestion
listener if not explicitly provided. We recommend that you
align the port numbers on as many nodes as possible to simplify the deployment
```
